lib/winapi/sync.c

```diff
@@ -38,6 +40,141 @@ VOID LeaveCriticalSection (LPCRITICAL_SECTION lpCriticalSection)
+        if (((ULONG_PTR)lpInitOnce->Ptr & INITONCE_MASK) != INITONCE_DONE) {
+            *fPending = TRUE;
+            SetLastError(ERROR_GEN_FAILURE);
+            return FALSE;
+        }
+
+        if (lpContext) {
+            *lpContext = (LPVOID)((ULONG_PTR)lpInitOnce & ~INITONCE_MASK);
+        }
+        // success
+        *fPending = FALSE;
+        return TRUE;
+    }
+
+    while (true) {
+        switch (((ULONG_PTR)lpInitOnce->Ptr) & INITONCE_MASK) {
+            case INITONCE_UNINITIALIZED:
+                if (__sync_bool_compare_and_swap(&lpInitOnce->Ptr, (PVOID)INITONCE_UNINITIALIZED, (dwFlags & INIT_ONCE_ASYNC) ? (PVOID)INITONCE_ASYNC_IN_PROGRESS : (PVOID)INITONCE_IN_PROGRESS)) {
+                    *fPending = TRUE;
+                    return TRUE;
+                }
+                break;
+            case INITONCE_IN_PROGRESS:
+                if (dwFlags & INIT_ONCE_ASYNC) {
+                    SetLastError(ERROR_INVALID_PARAMETER);
+                    *fPending = TRUE;
+                    return FALSE;
+                }
+
+                SwitchToThread();
+                __sync_synchronize();
+                break;
+            case INITONCE_ASYNC_IN_PROGRESS:
+                if (!(dwFlags & INIT_ONCE_ASYNC)) {
+                    SetLastError(ERROR_INVALID_PARAMETER);
+                    *fPending = TRUE;
+                    return FALSE;
+                }
+                *fPending = TRUE;
+                return TRUE;
+            case INITONCE_DONE:
+                if (lpContext) {
+                    *lpContext = (PVOID)((ULONG_PTR)lpInitOnce->Ptr & ~INITONCE_MASK);
+                }
+                *fPending = FALSE;
+                return TRUE;
+                break;
+        }
+    }
+}
+
+BOOL InitOnceExecuteOnce (PINIT_ONCE InitOnce, PINIT_ONCE_FN InitFn, PVOID Parameter, LPVOID *Context)
+{
+    if ((((ULONG_PTR)Context) & INITONCE_MASK) != INITONCE_UNINITIALIZED) {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+
+    BOOL fPending;
+    BOOL ret;
+    ret = InitOnceBeginInitialize(InitOnce, 0, &fPending, Context);
+    if (!fPending) {
+        return ret;
+    }
+
+    if (InitFn(InitOnce, Parameter, Context)) {
+        return InitOnceComplete(InitOnce, 0, Context ? *Context : NULL);
+    }
+
+    InitOnceComplete(InitOnce, INIT_ONCE_INIT_FAILED, NULL);
+    return FALSE;
+}
+
+BOOL InitOnceComplete (LPINIT_ONCE lpInitOnce, DWORD dwFlags, LPVOID lpContext)
+{
+    if (((ULONG_PTR)lpContext & INITONCE_MASK) != INITONCE_UNINITIALIZED) {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+
+    if (dwFlags & INIT_ONCE_INIT_FAILED) {
+        if (lpContext || (dwFlags & INIT_ONCE_ASYNC)) {
+            SetLastError(ERROR_INVALID_PARAMETER);
+            return FALSE;
+        }
+    } else {
+        lpContext = (PVOID)((ULONG_PTR)lpContext | INITONCE_DONE);
+    }
+
+    while (true) {
+        void *cur_val = lpInitOnce->Ptr;
```

Comment:
nit: For clarity, `PVOID` would be good as type.

The name `lpInitOnce->Ptr` suggested to me that taking the pointer to it was wrong in `__sync_val_compare_and_swap`, but it appears to be correct?
If this was `PVOID`, then it would be more obvious that the first argument has to be `PVOID*` (so the `&` below is correct, as the intention is to change where the pointer points / state bits in pointer; not change data where the pointer points).